sdk/identity/src/token_credentials/powershell_credential.rs

```diff
@@ -0,0 +1,190 @@
+
+lazy_static::lazy_static! {
+    static ref POWERSHELL_PATH: PathBuf = get_default_powershell_path();
+}
+
+fn get_scope_resource(scope: &str) -> String {
+    Regex::new(r"/.default$")
+        .unwrap()
+        .replace(scope, "")
+        .to_string()
+}
+
+#[cfg(target_os = "windows")]
+fn get_default_powershell_path() -> PathBuf {
+    match Command::new("pwsh.exe").args("/?").output() {
+        Ok(_) => PathBuf::from_str("pwsh.exe"),
+        Err(_) => PathBuf::from_str("powershell.exe"),
+    }
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_default_powershell_path() -> PathBuf {
+    PathBuf::from_str("pwsh").unwrap()
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+struct PowerShellTokenResponse {
+    pub token: AccessToken,
+    pub expires_on: DateTime<Utc>,
+}
+
+/// Enables authenticating with Azure Active Directory using Azure PowerShell.
+///
+/// To use this credential:
+/// - Install the Azure PowerShell Module
+///   `Install-Module -Name Az -Scope CurrentUser -Repository PSGallery -Force`
+/// - Ensure that you are signed in using the `Connect-AzAccount` cmdlet.
+pub struct AzurePowerShellCredential {
+    powershell_path: Option<PathBuf>,
+}
+
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error)]
+pub enum AzurePowerShellCredentialError {
+    #[error("failed to import the Az.Account module (is it installed?): {0}")]
+    ImportFailed(String),
+    #[error("failed to authenticate using the Az.Account module (are you logged in using `Connect-AzAccount`?): {0}")]
+    CommandFailed(String),
+    #[error("unable to execute PowerShell: {0}")]
+    PowerShellNotExecutable(Error),
+    #[error("failed to receive PowerShell subprocess output: {0}")]
+    ReceiveFailed(Canceled),
+    #[error("the token response from Az.Account was not UTF-8 encoded")]
+    ResponseNotUtf8(Utf8Error),
+    #[error("failed to deserialize Az.Account token response")]
+    ResponseFailedToDeserialize(serde_json::Error),
+    #[error("unknown error of kind: {0:?}")]
+    UnknownError(ErrorKind),
+}
+
+impl From<Error> for AzurePowerShellCredentialError {
+    fn from(error: Error) -> Self {
+        match error.kind() {
+            ErrorKind::NotFound => AzurePowerShellCredentialError::PowerShellNotExecutable(error),
+            error_kind => AzurePowerShellCredentialError::UnknownError(error_kind),
+        }
+    }
+}
+
+impl AzurePowerShellCredential {
+    pub fn new(powershell_path: Option<PathBuf>) -> Self {
+        Self {
+            powershell_path: powershell_path,
+        }
+    }
+
+    /// A helper to run PowerShell commands asynchronously using threads and an async_channel
+    async fn execute_powershell_command<I, S>(
+        &self,
+        args_iter: I,
+    ) -> Result<Output, AzurePowerShellCredentialError>
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<OsString>,
+    {
+        let pwsh = self.powershell_path.as_ref().unwrap_or(&*POWERSHELL_PATH);
```

Comment:
Won't this cause `get_default_powershell_path` to run on the current thread? I think we want this also to run on a separate thread no?

Reply:
Also, you'll want to use `unwrap_or_else` so that `POWERSHELL_PATH` is only evaluated if `self.powershell_path` is `None`.